libpermian/plugins/anaconda_webui/__init__.py

```diff
@@ -0,0 +1,443 @@
+        
+        self.log('Running: ' + ' '.join(cmd))
+        if self.test_repo_name:
+            # Test in separate repo
+            test_env = {'WEBUI_TEST_DIR': os.path.abspath(os.path.join(self.webui_dir, 'test'))}
+            if self.use_container:
+                cont_cwd = self.test_repo_dir.replace(self.temp_dir, '/root/workdir')
+                test_env['WEBUI_TEST_DIR'] = test_env['WEBUI_TEST_DIR'].replace(self.temp_dir, '/root/workdir')
+                process = self.container.exec(cmd, self.temp_dir, env=test_env, cwd=cont_cwd)
+            else:
+                process = subprocess.run(cmd, cwd=self.test_repo_dir, env=test_env, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+
+        else:
+            # Test in anaconda repo
+            if self.use_container:
+                cont_cwd = self.anaconda_dir.replace(self.temp_dir, '/root/workdir')
+                process = self._run_in_container(cmd, self.temp_dir, cwd=cont_cwd)
+            else:
+                process = subprocess.run(cmd, cwd=self.anaconda_dir, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+
+        self.log(process.stdout.decode(), 'output')
+        self.log('Test finished', show=True)
+
+        if process.returncode != 0:
+            self.reportResult(Result('complete', 'FAIL', True))
+        else:
+            self.reportResult(Result('complete', 'PASS', True))            
+
+    def teardown(self):
+        self.log('Teardown', show=True)
+        # Kill VM        
+        self._virsh_call(['destroy', self.vm_name])
+        # Remove VM
+        self._virsh_call(['undefine', self.vm_name, '--remove-all-storage'], check=True)
+        
+        self.vm_semaphore.release()
+
+        with self.setup_lock:
+            # If this instance is the last one using temp dir -> remove it
+            self.instances[self.git_anaconda_branch].remove(self)
+            if len(self.instances[self.git_anaconda_branch]) == 0:
+                self.temp_dirs[self.git_anaconda_branch].cleanup()
+                
+                del self.instances[self.git_anaconda_branch]
+                # Remove container image, if this was the last anacoda-webui workflow
+                if self.instances == dict():
+                    self.container.remove_image()
+
+    def terminate(self):
+        # Kill VM
+        self._virsh_call(['destroy', self.vm_name])
+
+    def log(self, message, name='workflow', show=False):
+        if name == 'workflow' and show:
+            self.last_log = message
+        super().log(message, name)
+
+    def displayStatus(self):
+        return self.last_log
+
+    def _start_vm(self):
+        """Prepare virt-install command, start VM and wait for IP address
+
+        :return: IP address
+        :rtype: str
+        """
+        self.log('Starting VM', show=True)
+
+        # Get compose location
+        os_url = self.installation_source.repos[self.installation_source.base_repo][self.architecture]
```

Comment:
The `installation_source.repos[self.installation_source.base_repo][self.architecture]` could be exposed by the structure e.g. like `installation_source.base_repo(self.architecture)`. This would of course need to have the original `base_repo` renamed to something like `base_repo_id` which would also make the structure more consistent.

Reply:
The `installation_source.base_repo[self.architecture]['os']` sounds also good.